### PR TITLE
TEST/IODEMO: fixed assert

### PR DESCRIPTION
**Pull request by hoopoepg**
- fixed assert: io_demo.cc:1140: size_t DemoClient::get_server_index(const UcxConnection*):
  Assertion `_server_index_lookup.size() == _active_servers.size()' failed

### CODE REVIEW

**Review by dmitrygx — APPROVED** [review]

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -1366,16 +1366,17 @@ class DemoClient : public P2pDemoCommon {
         LOG << "terminate connection " << server_info.conn << " due to "
             << reason;
 
+        // Don't wait for any more completions on this connection
+        _num_sent -= get_num_uncompleted(server_info);
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1370]: need to make sure any completions which may arrive during the destructor progress will not be counted
maybe assert/check that "conn" is member of _server_index_lookup map when counting completion?
@evgeny-leksikov WDYT?

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:1370]: I think simpler and clearer to check counting with `assert(_num_completed <= _num_sent) ` and `get_num_uncompleted(server_info) <= get_num_completed(server_info) ` in `DemoClient::handle_operation_completion()`, check in lookup map could be additional to make sure we handle structures correctly

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc:1376]: ok, added asserts

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -1335,6 +1338,9 @@ class DemoClient : public P2pDemoCommon {
     }
 
     long get_num_uncompleted(const server_info_t& server_info) const {
+        assert(server_info.num_sent >= (server_info.num_completed[IO_READ] +
+                                        server_info.num_completed[IO_WRITE]));
+
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1343]: maybe calculate the result (as long) and assert it's >= 0 ?

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -1366,16 +1372,17 @@ class DemoClient : public P2pDemoCommon {
...
             << reason;
 
+        // Don't wait for any more completions on this connection
+        _num_sent -= get_num_uncompleted(server_info);
+
         // Remove connection pointer
         _server_index_lookup.erase(server_info.conn);
 
+        // Remove active servers entry
+        active_servers_remove(server_info.active_index);
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1382]: maybe better do `delete server_info.conn` first, and only then `_server_index_lookup.erase(server_info.conn);` ?
@hoopoepg can you pls post the full backtrace of the assertion?

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:1343]: `long` can't protect from negative overflow, it's always >= 0
I implemented the same a bit different way in other PR https://github.com/openucx/ucx/pull/6338/files#diff-a12f7a78e9bc511657410e040a48b180f47969fbfcf2738f76136249fb13244fR1351

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc:1382]: I can't: core files were deleted and no new runs on last week.
crash stack is looks like close_server->ucp_ep_close_nb + ucp_worker_progress->dispatch_io_message->assert

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:1343]: isn't `long` signed?

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: updated assert

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: > isn't `long` signed?

you are right

**Comment by yosefe** [comment]
relevant:
```
2021-03-09T12:15:45.9301108Z [command]/usr/bin/bash --noprofile --norc /labhome/swx-azure-svc/workspace/azure/io_demo_sputnik/_temp/7e41414f-87ad-4fb1-b676-65162de4a013.sh
2021-03-09T12:15:46.0580909Z Error iodemo analyzer: contains error: io_demo: io_demo.cc:1175: void DemoClient::handle_operation_completion(size_t, io_op_t): Assertion `_num_completed < _num_sent' failed.
2021-03-09T12:15:46.0582306Z 
2021-03-09T12:15:46.0582899Z Log drop_14437/tag/iodemo_sputnik2_client_00.log:
2021-03-09T12:15:46.0586970Z Line io_demo: io_demo.cc:1175: void DemoClient::handle_operation_completion(size_t, io_op_t): Assertion `_num_completed < _num_sent' failed.
2021-03-09T12:15:46.0588290Z 
2021-03-09T12:15:46.0588555Z 
2021-03-09T12:15:46.0589137Z Traceback (most recent call last):
2021-03-09T12:15:46.0589943Z   File "drop_14437/buildlib/io_demo/iodemo_analyzer.py", line 229, in <module>
2021-03-09T12:15:46.0590837Z     process_client(clients, args.no_traffic_in_sec, args.seek, args.allow_list)
2021-03-09T12:15:46.0591575Z   File "drop_14437/buildlib/io_demo/iodemo_analyzer.py", line 191, in process_client
2021-03-09T12:15:46.0592352Z     raise Exception("contains error: {}\nLog {}:\nLine {}".format(line, log, line))
2021-03-09T12:15:46.0593874Z Exception: contains error: io_demo: io_demo.cc:1175: void DemoClient::handle_operation_completion(size_t, io_op_t): Assertion `_num_completed < _num_sent' failed.
2021-03-09T12:15:46.0594665Z 
2021-03-09T12:15:46.0595140Z Log drop_14437/tag/iodemo_sputnik2_client_00.log:
2021-03-09T12:15:46.0597032Z Line io_demo: io_demo.cc:1175: void DemoClient::handle_operation_completion(size_t, io_op_t): Assertion `_num_completed < _num_sent' failed.
2021-03-09T12:15:46.0597834Z 
2021-03-09T12:15:46.0700306Z ##[error]Bash exited with code '1'.
2021-03-09T12:15:46.0721000Z ##[section]Finishing: Analyze for tag
```

File: test/apps/iodemo/io_demo.cc (hunk 1 above)
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:1379]: move `delete server_info.conn` before this line

**Review by yosefe — APPROVED** [review]

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -1366,16 +1375,20 @@ class DemoClient : public P2pDemoCommon {
...
-
         // Destroying the connection will complete its outstanding operations
         delete server_info.conn;
 
         // Don't wait for any more completions on this connection
         _num_sent -= get_num_uncompleted(server_info);
 
+        // Remove connection pointer
+        // server_info.conn is used as ptr value, no real access to object
+        /* coverity[use_after_free] */
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1386]: better avoid coverity comments in example code

Comment (hoopoepg) [reply test/apps/iodemo/io_demo.cc]: ok, re-implemented

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@hoopoepg pls squash

**Comment by hoopoepg** [comment]
infra issue

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Pull request merged**